### PR TITLE
Return correct Bart hidden state tensors

### LINKED ISSUE #8601: Accessing gradients of Bart hidden states
The forums suggested that this be filed as a bug report:

https://discuss.huggingface.co/t/finding-gradients-in-zero-shot-learning/2033/5

The solution to the problem was solved on SO:

https://stackoverflow.com/questions/64823332/gradients-returning-none-in-huggingface-module/64866990#64866990

The question and answer are reproduced below. Filling as an issue as we should be able to compute gradients on output without a monkey-patch. It looks like the `transpose` is causing it.

## Environment info

- `transformers` version: 3.4.0
- Platform: Linux-4.15.0-123-generic-x86_64-with-glibc2.27
- Python version: 3.8.1
- PyTorch version (GPU?): 1.7.0 (True)
- Tensorflow version (GPU?): not installed (NA)
- Using GPU in script?: CPU & GPU
- Using distributed or parallel set-up in script?: No

### Who can help

 Bart: @patrickvonplaten

## Information

Model I am using (Bert, XLNet ...):

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [x] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [x] my own task or dataset: (give details below)

## To reproduce

```python
from transformers import pipeline
import torch

model_name = 'facebook/bart-large-mnli'
nlp = pipeline("zero-shot-classification", model=model_name)

responses = ["I'm having a great day!!"]
hypothesis_template = 'This person feels {}'
candidate_labels = ['happy', 'sad']
nlp(responses, candidate_labels, hypothesis_template=hypothesis_template)
```

This works well! The output is:

```
{'sequence': "I'm having a great day!!",
 'labels': ['happy', 'sad'],
 'scores': [0.9989933371543884, 0.0010066736722365022]}
```

What I'd like to do however, is look at the gradients of the input tokens to see which tokens are important. This is in contrast to looking at the attention heads (which is also another viable tactic). Trying to rip apart the internals of the module, I can get the logics and embedding layers:

```
inputs = nlp._parse_and_tokenize(responses, candidate_labels, hypothesis_template)
predictions = nlp.model(**inputs, return_dict=True, output_hidden_states=True)
predictions['logits']

tensor([[-3.1864, -0.0714,  3.2625],
        [ 4.5919, -1.9473, -3.6376]], grad_fn=<AddmmBackward>)
```

This is expected, as the label for "happy" is index 0 and the entailment index for this model is 2, so the value of 3.2625 is an extremely strong signal. The label for "sad" is 1 and the contradiction index is 0, so the value of 4.5919 is also the correct answer. 

Great! Now I should be able to look at the first embedding layer and check out the gradient with respect to the happy entailment scalar:

```
layer = predictions['encoder_hidden_states'][0]
layer.retain_grad()
predictions['logits'][0][2].backward(retain_graph=True)
```

Unfortunately, `layer.grad` is `None`. 

## [Solution from StackOverflow](https://stackoverflow.com/a/64866990/249341)

I was also very surprised of this issue. Although I have never used the library I went down and did some debugging and found out that the issue is coming from the library transformers. The problem is comming from from this [line][1] :

    encoder_states = tuple(hidden_state.transpose(0, 1) for hidden_state in encoder_states)

If you comment it out, you will get the gradient just with some dimensions transposed.
This issue is related to the fact that Pytorch Autograd does not do very well on inplace operations as mentioned [here][2].

So to recap the solution is to comment line 382 in *`modeling_bart.py`*.

You will get the gradient with this shape T x B x C instead of  B x T x C, but you can reshape it as you want later.

  [1]: https://github.com/huggingface/transformers/blob/1073a2bde5d608f9891d6da6df7b63921dca1b71/src/transformers/modeling_bart.py#L382
  [2]: https://discuss.pytorch.org/t/encounter-the-runtimeerror-one-of-the-variables-needed-for-gradient-computation-has-been-modified-by-an-inplace-operation/836/5

### PR DESCRIPTION
**Pull request by joeddav**
# What does this PR do?

Fixes #8601. When `output_hidden_states=True`, hidden states transposing is done _before_ being fed through the next layer. This ensures that the returned hidden state tensors lie upstream in the graph from the model outputs (allowing their gradients to be computed).

### CODE REVIEW

**Review by patrickvonplaten — COMMENTED** [review]
Your PR is great - `torch.transpose` does not allocate new memory, so it's very cheap! Great!

Super interesting this issue in general - I took some time to play around with it. I think for our researcher community it'd actually be an important feature to have for all models.

It would be great if you could take this PR as an opportunity to make sure this feature is enabled for `output_attentions` and `output_hidden_states` for all models in the lib. I would propose to add a test to `test_modeling_common.py` that looks as follows:

```python
    def test_retain_grad_hidden_states_attentions(self):
        config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
        config.output_hidden_states = True
        config.output_attentions = True

        # no need to test all models as different heads yield the same functionality
        model_class = self.all_model_classes[0]
        model = model_class(config)
        model.to(torch_device)

        inputs = self._prepare_for_class(inputs_dict, model_class)

        outputs = model(**inputs)
        first_output = outputs[0]

        if not config.is_encoder_decoder:
            first_hidden_states = outputs.hidden_states[0]
            first_attentions = outputs.attentions[0]

            first_hidden_states.retain_grad()
            first_attentions.retain_grad()

            first_output.flatten()[0].backward(retain_graph=True)

            self.assertIsNotNone(first_hidden_states.grad)
            self.assertIsNotNone(first_attentions.grad)

        else:
            first_encoder_hidden_states = outputs.encoder_hidden_states[0]
            first_encoder_attentions = outputs.encoder_attentions[0]
            first_encoder_hidden_states.retain_grad()
            first_encoder_attentions.retain_grad()

            first_decoder_hidden_states = outputs.decoder_hidden_states[0]
            first_decoder_attentions = outputs.decoder_attentions[0]
            first_decoder_hidden_states.retain_grad()
            first_decoder_attentions.retain_grad()

            first_output.flatten()[0].backward(retain_graph=True)

            self.assertIsNotNone(first_encoder_hidden_states.grad)
            self.assertIsNotNone(first_encoder_attentions.grad)
            self.assertIsNotNone(first_decoder_hidden_states.grad)
            self.assertIsNotNone(first_decoder_attentions.grad)
```

If Bart has this problem it is very likely that `FSTM` and `ProphetNet` have it as well actually...so this test would show this. 

Also we would have to do the same trick for the attentions in Bart, *e.g.* we could do the following change: 

```python
        if output_attentions:
            attn_weights_output = attn_weights.view(bsz, self.num_heads, tgt_len, src_len)
            attn_weights = attn_weights_output.view(bsz * self.num_heads, tgt_len, src_len)
        else:
            attn_weights_output = None

        attn_probs = F.dropout(attn_weights, p=self.dropout, training=self.training)

        assert v is not None
        attn_output = torch.bmm(attn_probs, v)
        assert attn_output.size() == (bsz * self.num_heads, tgt_len, self.head_dim)
        attn_output = attn_output.transpose(0, 1).contiguous().view(tgt_len, bsz, embed_dim)
        attn_output = self.out_proj(attn_output)

        return attn_output, attn_weights_output
```

in the `Attention` class of Bart.

Some models will always fail this test - I'd assume Reformer, Longformer, but also ProphetNet decoder to fail this test. For these models we can just overwrite the test in the model's specific test class - either completely skip it or run parts of it.

@joeddav - I'm happy to help on this - let me know :-)

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 1]
```diff
@@ -706,7 +706,7 @@ def forward(
                 if isinstance(head_mask, torch.Tensor):
                     head_mask = head_mask.to(hidden_states.device)
             if output_hidden_states:
-                all_hidden_states = all_hidden_states + (hidden_states.view(*output_shape),)
```
Comment (patrickvonplaten) [inline src/transformers/models/gpt2/modeling_gpt2.py:709]: this is not needed as `output_shape` is always equal to `hidden_states.shape`

File: src/transformers/models/openai/modeling_openai.py [hunk 2]
```diff
@@ -502,7 +502,7 @@ def forward(
         all_hidden_states = () if output_hidden_states else None
         for i, block in enumerate(self.h):
             if output_hidden_states:
-                all_hidden_states = all_hidden_states + (hidden_states.view(*output_shape),)
```
Comment (patrickvonplaten) [inline src/transformers/models/openai/modeling_openai.py:505]: same here -> `output_shape` is always equal to `hidden_states.shape`

**Comment by patrickvonplaten** [comment]
## UPDATE:

This PR is ready for review.

@sgugger, @LysandreJik - this very-well documented issue: https://github.com/huggingface/transformers/issues/8601 shows that for some models the gradient of the outputted `hidden_states` and the `attentions` cannot be computed because the tensors are excluded from the computation graph via some `transpose`, `permute`, or `slice` operations. 
@joeddav found a great fix for Bart and I applied the same fix now for all other models and added a test.

The only models that are not capable of keeping the gradient in `attentions` and `hidden_states` are 
- Longformer -> chunked attention slice operations don't allow keeping grad
- Reformer -> customized  backward doesn't help here
- ProphetNet -> Decoder part can't keep grad because of slice operations
- TransfoXL, XLNet -> two stream attention doesn't allow to keep the grad either

All other models can keep the grad which is ensured by the test.

**Review by patrickvonplaten — APPROVED** [review]

**Review by sgugger — APPROVED** [review]
I'm not expert enough in the internals of those models so I'll trust both of you on this. Code is very clean :-)

**Review by joeddav — COMMENTED** [review]
Lgtm, thanks for the help!

File: src/transformers/models/fsmt/modeling_fsmt.py [hunk 3]
```diff
@@ -666,7 +666,9 @@ def forward(
         for idx, decoder_layer in enumerate(self.layers):
             # add LayerDrop (see https://arxiv.org/abs/1909.11556 for description)
             if output_hidden_states:
+                x = x.transpose(0, 1)
                 all_hidden_states += (x,)
+                x = x.transpose(0, 1)
```
Comment (LysandreJik) [inline src/transformers/models/fsmt/modeling_fsmt.py:671]: Is it possible to display the information about dimension changes here too? I think those comments are really helpful in the understanding of what's happening here.

File: tests/test_modeling_common.py [hunk 4]
```diff
@@ -689,6 +689,56 @@ def check_hidden_states_output(inputs_dict, config, model_class):
 
             check_hidden_states_output(inputs_dict, config, model_class)
 
+    def test_retain_grad_hidden_states_attentions(self):
```
Comment (LysandreJik) [inline tests/test_modeling_common.py:692]: Very nice test.

File: tests/test_modeling_lxmert.py [hunk 5]
```diff
@@ -697,3 +697,36 @@ def check_hidden_states_output(inputs_dict, config, model_class):
             config.output_hidden_states = True
 
             check_hidden_states_output(inputs_dict, config, model_class)
+
+    def test_retain_grad_hidden_states_attentions(self):
```
Comment (LysandreJik) [inline tests/test_modeling_lxmert.py:701]: Thank you for implementing it for LXMERT as well!

**Review by LysandreJik — APPROVED** [review]
LGTM, this is great!
